### PR TITLE
Gird

### LINKED ISSUE #572: GroupBy via plywood
Currently when we're generating query in https://github.com/allegro/turnilo/blob/master/src/common/utils/query/visualization-query.ts we get nested set of topN's. Thats because we allow for custom sorts and splits on splits. It is not possible to generate group-by query with different sorts on different split dimensions. Probably it is not possible to get top N values in respective splits. 

We should check with what constraints we can generate group-by query. 

That way we could generate faster queries for few common cases and for example let user pick bigger limits. Common split values would also let us create stacked bar charts/mosaic plots. 

If that would be possible we need to discuss how present this option to the user. I don't think it should be transparent because changing sort could change limit settings behind user back. Probably different visualisations? Maybe different mode altogether?

### PR DESCRIPTION
**Pull request by adrianmroz-allegro**

### CODE REVIEW

File: src/client/modals/druid-query-modal/druid-query-modal.tsx [hunk 1]
```diff
@@ -29,8 +30,9 @@ interface DruidQueryModalProps {
 }
 
 export const DruidQueryModal: React.SFC<DruidQueryModalProps> = ({ onClose, timekeeper, essence }) => {
-  const { dataCube: { attributes, source, options: { customAggregations, customTransforms } } } = essence;
-  const query = makeQuery(essence, timekeeper);
+  const { visualization, dataCube: { attributes, source, options: { customAggregations, customTransforms } } } = essence;
+  const queryFn = visualization.name === "grid" ? gridQuery : standardQuery;
```
Comment (mkuthan) [inline src/client/modals/druid-query-modal/druid-query-modal.tsx:34]: leaking abstraction, visualisation is responsible for "queryfn" so could be move this logic into visualisation itself?

File: src/client/visualizations/base-visualization/base-visualization.tsx [hunk 2]
```diff
@@ -109,8 +109,12 @@ export class BaseVisualization<S extends BaseVisualizationState> extends React.C
     return this.debouncedCallExecutor(essence, timekeeper);
   }
 
+  protected getQuery(essence: Essence, timekeeper: Timekeeper): Expression {
```
Comment (mkuthan) [inline src/client/visualizations/base-visualization/base-visualization.tsx:112]: why this function only delegates to another function?

File: src/client/visualizations/grid/grid.tsx [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      segmentWidth: SEGMENT_WIDTH,
+      ...super.getDefaultState()
+    };
+  }
+
+  setScroll = (scrollTop: number, scrollLeft: number) => this.setState({ scrollLeft, scrollTop });
+
+  setSegmentWidth = (segmentWidth: number) => this.setState({ segmentWidth });
+
+  private getIdealColumnWidth(): number {
```
Comment (mkuthan) [inline src/client/visualizations/grid/grid.tsx:62]: How to test calculation if method is private? I would extract layout math form the component class.

File: src/client/visualizations/table/body/measures/measure-row.tsx [hunk 4]
```diff
@@ -50,7 +49,6 @@ export const MeasureRow: React.SFC<MeasureRowProps> = props => {
         highlight={highlight}
         scale={scales[i]}
         cellWidth={cellWidth}
-        lastLevel={datum["__nest"] === splitLength}
```
Comment (mkuthan) [inline src/client/visualizations/table/body/measures/measure-row.tsx:53]: What is that, regression?

File: src/common/visualization-manifests/grid/grid.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    const { sort: firstSort } = splits.getSplit(0);
+    const sort = firstSort instanceof SeriesSort
+      ? firstSort
+      : new SeriesSort({ reference: firstSeries.reference });
+    const newSplits = splits.update("splits", splits =>
+      splits.map(split =>
+        split.changeLimit(100).changeSort(sort)));
+
+    if (splits.equals(newSplits)) {
+        return Resolve.ready(isSelectedVisualization ? 10 : 6);
```
Comment (mkuthan) [inline src/common/visualization-manifests/grid/grid.ts:39]: Magic numbers, do we have constants for that?

**Review by mkuthan — COMMENTED** [review]
Looks good - small impact on existing visualisations, especially for query assembly which is different for grid.
A few questions/comments raised.

File: src/client/modals/druid-query-modal/druid-query-modal.tsx (hunk 1 above)
Comment (adrianmroz) [reply src/client/modals/druid-query-modal/druid-query-modal.tsx:34]: Yes, very leaking. These components are very far apart and don't have a good point to communicate. I knowingly made this change in the name of technical debt :)

File: src/client/visualizations/base-visualization/base-visualization.tsx (hunk 2 above)
Comment (adrianmroz) [reply src/client/visualizations/base-visualization/base-visualization.tsx:112]: This is base/default implementation. Specific Visualisations can override this.

And logic is extracted to function for better maintainability.

File: src/client/visualizations/table/body/measures/measure-row.tsx (hunk 4 above)
Comment (adrianmroz) [reply src/client/visualizations/table/body/measures/measure-row.tsx:53]: No, this logic is extracted higher in component tree and better named.

File: src/common/visualization-manifests/grid/grid.ts (hunk 5 above)
Comment (adrianmroz) [reply src/common/visualization-manifests/grid/grid.ts:39]: We don't and honestly I don't know what specific numbers mean. I think we need to heavily rework this mechanism in future.

File: src/client/visualizations/grid/grid.tsx (hunk 3 above)
Comment (adrianmroz) [reply src/client/visualizations/grid/grid.tsx:62]: Probably you're right. As always, these functions mostly connect multiple variables from object state so tests will get messy. But extraction is good first step!

File: src/client/modals/druid-query-modal/druid-query-modal.tsx (hunk 1 above)
Comment (mkuthan) [reply src/client/modals/druid-query-modal/druid-query-modal.tsx:34]: sure, just make a note and we will discuss how to define visualisation-make-query abstraction later on

File: src/client/visualizations/base-visualization/base-visualization.tsx (hunk 2 above)
Comment (mkuthan) [reply src/client/visualizations/base-visualization/base-visualization.tsx:112]: clear, w need better abstraction for this in the future, just make a note for further work

File: src/client/visualizations/grid/make-query.ts [hunk 6]
```diff
@@ -40,7 +40,9 @@ function applySeries(series: List<ConcreteSeries>, timeShiftEnv: TimeShiftEnv, n
 }
 
 function applyLimit({ limit }: Split) {
-  const limitExpression = new LimitExpression({ value: limit });
+  // TODO: this calculation is for evaluation purpose. We should add custom split values for Grid and remove this multiplication!
+  const value = limit * 10;
```
Comment (mkuthan) [inline src/client/visualizations/grid/make-query.ts:44]: :)

**Review by mkuthan — APPROVED** [review]
Lets show beta and get some feedback!

**Pull request merged**